### ISSUE TITLE
in the separate state, it doesnt know about `~/.init.lua` either.

### ISSUE THREAD

**Issue by o-jasper** [luapower/luastate#1]
Note that this is coming from `thread`, and `pthread`, but probably the issue belongs here?

After `state:push(function()`, adding `loadfile(os.getenv("HOME") .. "/.init.lua")()` fixes in Linux. There is also `LUA_INIT`, though.. So i think...

```
local exec = os.getenv("LUA_INIT") or os.getenv("LUA_INIT_5_3")
local file = not exec and os.getenv("HOME") .. "/.init.lua" or string.match("^%*([.+])$")
if file then
    loadfile(file)()
else
    loadstring(exec)()
end
```

Seems like kindah a pita :/

**Comment by capr** [comment]
Hi Jasper,

I'm not sure I understand what the issue is. Is there a problem with env. vars in lua states? Please ellaborate. Thanks.

**Comment by o-jasper** [comment]
Basically some of the initialization of the lua state when you run, isnt done by the state object, i assume you use it plenty, so it is only a problem if you installed packages in such a way that it doesnt need initialization to access them. Initialization may make more packages available by adding to `package.path`.(afaik, only)

It seems to me at least that if you use `LUA_INIT`(the OS env var) or `~.init.lua`, then it is an issue.

If `~/.init.lua` is  `print("INIT")` This prints it once, it is called only by the luajit program, not state;

```
ls = require "luastate"
state = ls:open()

state:openlibs()

state:push(function() print("WORK") end)
state:call()
```

And this twice;

```
ls = require "luastate"
state = ls:open()

state:openlibs()

state:push(function() 
      local exec = os.getenv("LUA_INIT") or os.getenv("LUA_INIT_5_3")
      local file = not exec and os.getenv("HOME") .. "/.init.lua" or string.match("^%*([.+])$", exec)
      if file then
         loadfile(file)()
      else
         loadstring(exec)()
      end
      print("WORK")
end)
state:call()
```

Now i know a bit more, maybe what i am suggesting is to add:(roughly)

```
function State:openinit()
    self:dostring([[ 
          local exec = os.getenv("LUA_INIT") or os.getenv("LUA_INIT_5_3")
          local file = not exec and os.getenv("HOME") .. "/.init.lua" or string.match("^%*([.+])$", exec)
          if file then
             loadfile(file)()
          else
             loadstring(exec)()
          end
    ]])
end
```

So the lua state can be run identically as running it with the lua program, including initialzation.

To be honest, maybe you _dont want_ to have this variation in how things are installed, (adding to `package.path` vs putting in the way of the existing `package.path`, but as-is, there are no absolute paths _in userspace_ for `package.path` so you can only install things as root unless you change `package.path` first

(default `package.path` is `./?.lua;/usr/share/luajit-2.0.4/?.lua;/usr/local/share/lua/5.1/?.lua;/usr/local/share/lua/5.1/?/init.lua;/usr/share/lua/5.1/?.lua;/usr/share/lua/5.1/?/init.lua`, only `./?.lua` can be userspace, but then it depends on where the program is called from)

(sorry for making the comment so long..)

**Comment by o-jasper** [comment]
Alternatively, `state:setglobal` or something could be used to just set `package.path`?

**Comment by capr** [comment]
I think I understand now. So the problem is that when you make a Lua state, it doesn't look at LUA_INIT as the main interpreter does. But why do you think it should? That's a feature of the standalone interpreter not of the Lua library.

Update: I think I understand now why you need this -- basically you want require() to behave the same in a Lua state as it does in the main interpreter, i.e. execute `LUA_INIT` which changes package.path. And you can't change `LUA_PATH` instead because you need to set the path dynamically at runtime based on the location of the exe, is that right?

**Comment by capr** [comment]
Where do you set LUA_INIT? In a launcher script? In your ~/.bashrc?

**Comment by o-jasper** [comment]
Yep, that's exactly it. _Didnt know_ about `LUA_PATH`! Right.. doing 

```
LUA_INIT="" LUA_PATH="is a thing" luajit -e "print(package.path)"
```

... did not notice that. I remember trying stuff like that a while before though... Dunno why i didnt end up using it. In my defense, i dont see it in the man pages, it is in the manual though...

I use some lua code to generate the `package.path`, but is is merely some convenience stuff, i dont need it, so i can switch to `LUA_PATH`.. So maybe you could say that is sufficient.. Nevertheless might be good to have states use `package.path` from when the luastate is created, rather than the one of when the lua program started. (that way no new methods are needed, just maybe some more arguments at creation to choose the initial `package.path`)

In bash variables can be set with `export VAR=val`(in `~/.bashrc` or before the command `LUA_INIT="something" luajit $args`, they should be equivalent. I dont set it at all, i use the `~/init.lua` which lua and luajit both defaultly run initially.

**Comment by capr** [comment]
Well, you raise some good points. I would definitely be nice if require() could work the same in a Lua state as it does in the parent state. Inheriting package.[c]path is one step towards it. But it won't cover all the cases. Some users add stuff to package.loaders, and overriding require() itself is not unheard of either (and it's a legitimate thing to do sometimes). Also, you just made me realize that all the loaders in package.loaders need to be thread-safe (this affects [bundle](https://luapower.com/bundle) which adds loaders to load modules from embedded symbols in the executable). I think a general solution would require that any module which wants to change package.path or add loaders to package.loaders or override require() would have to be aware of luastate, which means they would have to register an "init hook" that would execute on every newly created state, which will prepare the state to make require() work seamlessly inside it.

**Comment by capr** [comment]
Thinking more about it, here's what I would do: #2. What do you think?

And while we're at it, how'bout this too? https://github.com/luapower/luajit/issues/2 (warning: you might not like this one)

Btw: I didn't know that Lua(JIT) executes `~/init.lua` by default, are you sure about that? I can't find anything in the docs about it.

**Comment by o-jasper** [comment]
Wow, i suck... apparently i added it via `LUA_INIT` `~/.bashrc` at some point and forgot about it.. Put a big comment ontop warning about it and need to revisit my readmes for references to this...

My first guess is that maybe if it gets too hairy, maybe it is better not to do it? Or try to keep the complexity separate.. At this point i don't know enough how it works internally though. Don't know if it is hairy. I was basically looking for multithreading so i could run a http server, and tox client(s) at the same time.

Do know that i.e. thinking about using `require` to try make "arbitrary lua files run on other servers" by having functions and tables replaced by objects that instead send http requests, you cannot deal with it if people mess with metatables. (Of course the ultimate Mandatory Access Control is if the thing not-to-be accessed isnt even on the server)

**Comment by capr** [comment]
Well, the implementation is trivial, and it's needed anyway if you want to use luastate and bundle together at least (there may be other uses too) so I think I'll implement this. Thanks for bringing it up.
